core/server/common/src/main/java/alluxio/master/journal/raft/DownloadObserver.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.journal.raft;
+
+import alluxio.grpc.DownloadSnapshotPRequest;
+import alluxio.grpc.DownloadSnapshotPResponse;
+import alluxio.grpc.SnapshotData;
+import alluxio.grpc.UploadSnapshotPRequest;
+import alluxio.grpc.UploadSnapshotPResponse;
+
+import io.grpc.stub.ClientCallStreamObserver;
+import io.grpc.stub.ClientResponseObserver;
+import io.grpc.stub.StreamObserver;
+import org.apache.ratis.io.MD5Hash;
+import org.apache.ratis.server.protocol.TermIndex;
+import org.apache.ratis.server.storage.FileInfo;
+import org.apache.ratis.statemachine.SnapshotInfo;
+import org.apache.ratis.statemachine.impl.SimpleStateMachineStorage;
+import org.apache.ratis.statemachine.impl.SingleFileSnapshotInfo;
+import org.apache.ratis.util.MD5FileUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+/**
+ * A stream observer for downloading a snapshot.
+ *
+ * @param <S> type of the message to send
+ * @param <R> type of the message to receive
+ */
+public class DownloadObserver<S, R> implements ClientResponseObserver<S, R> {
+  private static final Logger LOG = LoggerFactory.getLogger(DownloadObserver.class);
+
+  private final SimpleStateMachineStorage mStorage;
+  private final CompletableFuture<TermIndex> mFuture = new CompletableFuture<>();
+  private final Function<Long, S> mMessageBuilder;
+  private final Function<R, SnapshotData> mDataGetter;
+  private TermIndex mTermIndex;
+  private File mTempFile;
+  private FileOutputStream mOutputStream;
+  private long mBytesWritten = 0;
+  private StreamObserver<S> mStream;
+  private SnapshotInfo mSnapshotToInstall;
+
+  /***
+   * Builds a stream for leader to download a snapshot.
```

Comment:
Because the RPC service is only up on the leader/primary master, we provide two different RPC calls for handling snapshot download from leader and download from follower.

The leader version uses the message types defined for the RPC call `uploadSnapshot`. This is used for leader to download a snapshot from follower.

The follower version uses the message types defined for the RPC call `downloadSnapshot`. This is used for follower to download a snapshot from leader.